### PR TITLE
PHPORM-33 Add tests on Query\Builder methods

### PR DESCRIPTION
**Pull request by GromNaN**
Fix [PHPORM-33](https://jira.mongodb.org/browse/PHPORM-33)
Replace #4

- Add tests on query builder methods that don't need to be fixed.
- Throw exception when calling unsupported methods: `whereIntegerInRaw`, `orWhereIntegerInRaw`, `whereIntegerNotInRaw`, `orWhereIntegerNotInRaw`
- Throw an exception when `Query\Builder::where` is called with only a column name

### CODE REVIEW

File: tests/Query/BuilderTest.php [hunk 1]
```diff
@@ -54,16 +79,105 @@ public static function provideQueryBuilderToMql(): iterable
...
-        yield 'find in array' => [
+        /** @see DatabaseQueryBuilderTest::testBasicWhereIns */
+        yield 'whereIn' => [
             ['find' => [['foo' => ['$in' => ['bar', 'baz']]], []]],
             fn (Builder $builder) => $builder->whereIn('foo', ['bar', 'baz']),
         ];
 
+        yield 'whereIn associative array' => [
+            ['find' => [['id' => ['$in' => [45582, 2, 3]]], []]],
+            fn (Builder $builder) => $builder->whereIn('id', ['issue' => 45582, 'id' => 2, 3]),
```
[the remark this answers is not in the document]
Comment (jmikola) [reply tests/Query/BuilderTest.php]: Let's push this to a separate issue for now. I'm very much in favor of raising exceptions for poorly formatted values. As annoying as it may be for some users, I think it will ultimately avoid unexpected edge cases and bugs down the line.

File: tests/Query/BuilderTest.php [hunk 2]
```diff
@@ -247,6 +321,34 @@ public static function provideQueryBuilderToMql(): iterable
...
 
+        /** @see DatabaseQueryBuilderTest::testLimitsAndOffsets() */
+        yield 'offset limit' => [
+            ['find' => [[], ['skip' => 5, 'limit' => 10]]],
+            fn (Builder $builder) => $builder->offset(5)->limit(10),
+        ];
+
+        yield 'offset 0 limit 0' => [
+            ['find' => [[], []]],
+            fn (Builder $builder) => $builder->offset(0)->limit(0),
```
Comment (jmikola) [inline tests/Query/BuilderTest.php:332]: Offhand, do you know if zero is handled exactly the same as "null (reset)"?

If so, I'd be in favor of the following changes:

 * Rename this test to "offset limit zero (unset)"
 * Create a new test below named "offset limit null (unset)", based on this test. The point of these tests would be to demonstrate that zero and null values both result in the options being unspecified.
 * Create a new test here named "offset limit zero (reset)", based on the "offset limit null (reset)" test below. Like that, it would demonstrate that zero and null both reset the options back to an unset state

**Review by jmikola — CHANGES_REQUESTED** [review]

File: tests/Query/BuilderTest.php [hunk 3]
```diff
@@ -247,6 +321,34 @@ public static function provideQueryBuilderToMql(): iterable
...
+        ];
+
+        yield 'offset limit negative' => [
+            ['find' => [[], []]],
+            fn (Builder $builder) => $builder->offset(-5)->limit(-10),
```
[the remark this answers is not in the document]
Comment (GromNaN) [reply tests/Query/BuilderTest.php:337]: That's a Laravel feature. I'm not sure we should care about changing this behavior.
https://github.com/laravel/framework/blob/10.x/src/Illuminate/Database/Query/Builder.php#L2435-L2437

File: tests/TransactionTest.php [hunk 4]
```diff
@@ -332,7 +332,7 @@ public function testTransaction(): void
         $count = User::count();
         $this->assertEquals(2, $count);
 
-        $this->assertTrue(User::where('alcaeus')->exists());
+        $this->assertTrue(User::where('name', 'alcaeus')->exists());
```
Comment (GromNaN) [inline tests/TransactionTest.php:335]: Being too permissive didn't allow us to detect this incorrect query.

```diff
- { "alcaeus": null }
+ { "name": "alcaeus" }
```

The new exception made it surface.

File: src/Query/Builder.php [hunk 5]
```diff
@@ -923,6 +923,10 @@ public function where($column, $operator = null, $value = null, $boolean = 'and'
             }
         }
```
Comment (alcaeus) [inline src/Query/Builder.php:924]: Can't comment on the `if` itself, but I'm wondering if the condition should read `func_num_args() >= 3` - we wouldn't be removing the `$` from an operator if `$boolean` was defined as well 🤔

**Review by alcaeus — APPROVED** [review]
LGTM. Reminder to create a follow-up ticket for disallowing non-lists when using `whereIn`.

File: src/Query/Builder.php (hunk 5 above)
Comment (jmikola) [reply src/Query/Builder.php:924]: It's not clear to me how `$boolean` is used, but I see it's passed on to the parent method below. I think that's right, and this suggests another lingering bug.

I'll defer to you guys whether to fix here (if so, add a regression test that passes a fourth parameter to `where()` (I suppose an explicit `'and'` string is fine) or report a new JIRA issue and address in a subsequent PR. A new issue probably makes more sense for visibility in the release notes.

File: tests/Query/BuilderTest.php (hunk 2 above)
Comment (jmikola) [reply tests/Query/BuilderTest.php:337]: This comment might have been missed.

I'll defer to you if you'd like to make the suggested test changes. Otherwise, feel free to resolve.

**Review by jmikola — APPROVED** [review]
Two follow-up comments.

LGTM to merge as-is if you'd rather not make the test changes. In that case, the issue @alcaeus mentioned in https://github.com/GromNaN/laravel-mongodb/pull/14/files#r1273137525 should be tracked in a separate ticket.

File: src/Query/Builder.php (hunk 5 above)
Comment (GromNaN) [reply src/Query/Builder.php:924]: Fixed in https://github.com/GromNaN/laravel-mongodb/pull/20

File: tests/Query/BuilderTest.php (hunk 2 above)
Comment (GromNaN) [reply tests/Query/BuilderTest.php]: Test updated. Thanks for the reminder.
